Multi-dyno support via Redis

JasonEtco opened the pull request:
We've recently seen some scaling issues with Smee, and while that scaling is artificial, it made me realize that Smee doesn't actually work across dynos so increasing the number of running instances would actually cause problems. Because connections are kept in-memory, if clients can be connected to an instance that doesn't receive an event (because there's another instance that does).

So, this PR adds support for a Redis pub/sub message bus. When Smee gets an event, rather than calling `events.emit()`, it does `bus.emitEvent()`. This publishes to Redis, which every instance of Smee is subscribed to. When a message is received from Redis, we call `bus.emitLocalEvent()`, which calls `events.emit()`.

Note that I wanted Redis to be optional; folks should still feel free to deploy their own Smee _without_ Redis, so we only try to make a connection if `process.env.REDIS_URL` is configured.

I went a little overboard with this PR, as I was adding code I had to reorganize some stuff. TLDR, most of the work happens in `event-bus.js`. I move all the server files into `lib`.

While testing, I also added a `bin/send-payload` script:

```
bin/send-payload <channel>
```

Figured I'd keep it in here for future use.

- [x] New tests for `EventBus`
- [x] New docs for configuring Redis (and that it's optional)

gr2m commented on lib/redis-bus.js:
```diff
@@ -0,0 +1,59 @@
+    this.events = events
+
+    // If Redis isn't enabled, don't try to connect
+    if (process.env.REDIS_URL) {
```
Just a tiny nitpick, maybe a personal preference, but I prefer to return early instead of wrapping bigger chunks of code in an if statement

So I would do

```js
if (!process.env.REDIS_URL) return
```

and then continue without antother indentation

JasonEtco replied:
ooh good idea 👍

gr2m commented:
Is this still a draft? Looks solid to me, didn't test it though, I trust you did :)

JasonEtco commented:
Thanks @gr2m - it's still a draft, I want to add some tests and docs, but the implementation is unlikely to change much.

tcbyrd commented on lib/redis-bus.js:
```diff
@@ -0,0 +1,59 @@
+   * @param {string} channel - Channel name
+   * @param {any} payload
+   */
+  emitRemoteEvent ({ channel, payload }) {
```
It feels a bit odd to me that the method here is `emitRemoteEvent` but can fall back to `local`. Maybe we just rename this to `emitEvent` since it can be either remote or local depending on config?

```suggestion
  emitEvent ({ channel, payload }) {
```

JasonEtco replied:
That works, especially given the next comment (renaming to `EventBus` and abstracting the Redis inclusion entirely)

tcbyrd commented on lib/server.js:
```diff
@@ -6,29 +6,18 @@ const EventEmitter = require('events')
 const Raven = require('raven')
 
+const channelIsBanned = require('./channel-is-banned-middleware')
+const RedisBus = require('./redis-bus')
```
Since this can emit events to the in-memory emitter too, maybe we rename this to `EventBus`? Then we remove any concept at this level of "remote" vs "local". You just emit events, and if Redis is configured it just knows implicitly to use it under the hood. Or is that too "magical"?

JasonEtco replied:
I'm not opposed 👍 I think that since we don't use the internal event emitter in `lib/server.js`, it makes sense to abstract it properly to this new class. On that note, might even make sense to _define_ the `EventEmitter` on what is now `RedisBus` instead of in `lib/server.js` 🤔

tcbyrd replied:
> On that note, might even make sense to define the EventEmitter on what is now RedisBus instead of in lib/server.js

Yeh I was thinking along these lines as well.

JasonEtco replied:
Done 👍

JasonEtco commented on lib/redis-bus.js:
```diff
@@ -0,0 +1,59 @@
+  }
+
+  /**
+   * Emit an even to this machine's in-memory EventEmitter
```
Oooh thanks!

JasonEtco commented:
Alrighty, I've added some basic docs, tests and implemented @tcbyrd's suggestions for making the `EventBus` the main abstraction for managing events (independent of `REDIS_URL`). I'm going to try this out on staging before a production deploy, would love another 👀 from y'all @gr2m @tcbyrd 🙏

JasonEtco commented:
I've deployed this to production, setup Redis and enabled multiple dynos just to verify its all working correctly. So far, so good! I'm going to merge this now and continue to monitor it tomorrow - but if we're seeing missing payloads (which is the only thing I'm worried about here) then I think the only way we'll know about it is from users. In any case, merging!

SGrondin commented on lib/event-bus.js:
```diff
@@ -0,0 +1,61 @@
+    this.pub = new Redis(this.opts.connection)
+
+    // Subscribe to the Redis event channel
+    this.sub.subscribe(this.opts.namespace)
```
None of the subscribe() calls in this PR are passing a callback, so any error here will be lost. It could be a good idea to wrap all those calls into a promise and await them.

JasonEtco replied:
Thank you for taking a look here @sgrondin! Hm according to the type definitions of `ioredis`, `.subscribe` doesn't return a Promise 🤔 I'll do some testing there and try it out anyway.

SGrondin replied:
Yeah, the docs don't mention a promise, but I know from experience that it takes a callback as last argument. If it does return a promise then that's great. But if it doesn't you can just wrap the call in `new Promise((resolve, reject) => ...)` and `await` that.

SGrondin commented on lib/event-bus.js:
```diff
@@ -0,0 +1,61 @@
+      return this.emitLocalEvent(opts)
+    } else {
+      const message = JSON.stringify(opts)
+      return this.pub.publish(this.opts.namespace, message)
```
AFAIK publish() also takes a callback and without it errors are lost

JasonEtco replied:
This can definitely be `try/catch`ed as a Promise 👌

SGrondin commented on tests/event-bus.test.js:
```diff
@@ -0,0 +1,65 @@
+      expect(pubSpy.mock.calls[0]).toMatchSnapshot()
+
+      // Disconnect and cleanup
+      await bus.sub.quit()
```
👍

JasonEtco commented:
Thanks for jumping in here @SGrondin - that's a great point. I'm not sure what we'd do with Redis-related errors, but we should definitely be capturing them.

I'll do some more research, but if you're up for it a PR adding a basic addition (even just `.on('error')` would be ✨!
